### PR TITLE
ROCm decoding

### PR DESCRIPTION
**Pull request by cloudhan**

### CODE REVIEW

**Comment by tianleiwu** [comment]
Please fix pipeline errors.

File: onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu [hunk 1]
```diff
@@ -863,7 +863,7 @@ void KeyCacheExpansionKernelLauncher(const T* key_cache,
   equiv_head_size = (equiv_head_size & 1) == 0 ? (equiv_head_size >> 1) : equiv_head_size;
 
   // Here we know head_size is smaller than max_thread_num_per_block
-  int tpb = std::max(32, equiv_head_size);
+  int tpb = std::max(GPU_WARP_SIZE_HOST, equiv_head_size);
```
Comment (tianleiwu) [inline onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu:866]: why not use props.warpSize as well?

Comment (cloudhan) [reply onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu:866]: props is not available here, it is not passed in as the previous function.

File: onnxruntime/contrib_ops/cuda/transformers/beam_search.cc [hunk 2]
```diff
@@ -48,6 +46,10 @@ BeamSearch::BeamSearch(const OpKernelInfo& info)
                    GenerationCudaDeviceHelper::InitBeamState<float>,
                    GenerationCudaDeviceHelper::InitBeamState<MLFloat16>);
 
+#ifndef USE_ROCM
```
Comment (kunal-vaishnavi) [inline onnxruntime/contrib_ops/cuda/transformers/beam_search.cc:49]: Should this be `#ifdef USE_CUDA` instead? The `SetDeviceHelpers_Cuda` method in `onnxruntime/contrib_ops/cpu/transformers/beam_search.h` is using `#ifdef USE_CUDA`:

```
#ifdef USE_CUDA
  void SetDeviceHelpers_Cuda(
      const GenerationDeviceHelper::ReorderPastStateFunc& reorder_past_state_func,
      const GenerationDeviceHelper::InitCacheIndirFunc& init_cache_indir_func) {
    reorder_past_state_func_ = reorder_past_state_func;
    init_cache_indir_func_ = init_cache_indir_func;
  }
#endif
```

File: onnxruntime/contrib_ops/cuda/transformers/beam_search.cc [hunk 3]
```diff
@@ -48,6 +46,10 @@ BeamSearch::BeamSearch(const OpKernelInfo& info)
...
+  SetDeviceHelpers_Cuda(GenerationCudaDeviceHelper::ReorderPastState, GenerationCudaDeviceHelper::InitCacheIndir);
+#endif
```
Comment (kunal-vaishnavi) [inline onnxruntime/contrib_ops/cuda/transformers/beam_search.cc:51]: Should this be `#ifdef USE_CUDA` instead? The `SetDeviceHelpers_Cuda` method in `onnxruntime/contrib_ops/cpu/transformers/beam_search.h` is using `#ifdef USE_CUDA`:

```
#ifdef USE_CUDA
  void SetDeviceHelpers_Cuda(
      const GenerationDeviceHelper::ReorderPastStateFunc& reorder_past_state_func,
      const GenerationDeviceHelper::InitCacheIndirFunc& init_cache_indir_func) {
    reorder_past_state_func_ = reorder_past_state_func;
    init_cache_indir_func_ = init_cache_indir_func;
  }
#endif
```

**Comment by wangyems** [comment]
please note that the vectorized load is currently not used in ROCM ep here: https://github.com/microsoft/onnxruntime/blob/dabe710028f8b467d60dba6cad65147972f617bd/onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu#L877

File: onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu (hunk 1 above)
Comment (tianleiwu) [reply onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu:866]: Is it possible to pass props to be consistent?

File: onnxruntime/contrib_ops/cuda/transformers/beam_search.cc (hunk 3 above)
Comment (cloudhan) [reply onnxruntime/contrib_ops/cuda/transformers/beam_search.cc:51]: No, this should not be `#ifdef USE_CUDA`, the file exists in `.../cuda/...`, the only case it is not compiling for CUDA is that it is hipified.

For where it is `#ifdef USE_CUDA`, you should notice the file path is `.../cpu/...`

File: onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu (hunk 1 above)
Comment (cloudhan) [reply onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu:866]: @tianleiwu, I don't want to do it at the moment. The problem is the interface design. The `KeyCacheExpansionKernelLauncher` is called in `ExpandBuffer`, which exists both in cpu and cuda

https://github.com/microsoft/onnxruntime/blob/9206b7cdc61df811ace631c25428fdf7f1a1b687/onnxruntime/contrib_ops/cpu/transformers/generation_device_helper.h#L355-L363

https://github.com/microsoft/onnxruntime/blob/9206b7cdc61df811ace631c25428fdf7f1a1b687/onnxruntime/contrib_ops/cuda/transformers/generation_device_helper.h#L142-L150

Adding `cudaDeviceProp` parameter will somehow pollute the cpu interface...

**Review by tianleiwu — APPROVED** [review]

**Pull request merged**